### PR TITLE
feat: `flox pull|push` for managed envs

### PR DESCRIPTION
**Pull request by ysndr**
Implements `flox pull [--dir]` and `flox pull --remote [--dir]`.

Environment branches are updated using `fast-forward`s and fail if this is not possible.

### CODE REVIEW

**Comment by stahnma** [comment]
Can this be attached to an issue, put in the sprint and estimated? Also, moving out of review since it's marked draft.

**Comment by ysndr** [comment]
Currently missing tests as `git.hub.flox.dev` seems hardcoded at the moment and we are missing a test user to validate against (i think)

**Comment by ysndr** [comment]
Pull is implemented and mostly working.
I'm now working on some testing around it and check if I can add an integration test that speaks to floxhub.
Pulling is generally ready for review now

**Comment by ysndr** [comment]
Darwin check failing because we don't have an env floxtest/default branch for Intel mac on floxhub...

> #     fatal: Remote branch x86_64-darwin.default not found in upstream origin

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 1]
```diff
@@ -414,6 +412,33 @@ impl ManagedEnvironment {
     }
 }
 
+fn lock_env(
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:415]: nit nonblocking: this function name isn't super well differentiated from `lock_manifest`. Idk if there's something better, maybe `lock_generation`

File: crates/flox-rust-sdk/src/models/floxmetav2/mod.rs [hunk 2]
```diff
@@ -111,3 +155,84 @@ pub(super) fn floxmeta_dir(flox: &Flox, owner: &EnvironmentOwner) -> PathBuf {
...
+    use std::fs;
+    use std::str::FromStr;
+
+    use super::*;
+    use crate::flox::tests::flox_instance;
+    use crate::flox::EnvironmentName;
+    use crate::providers::git::GitProvider;
+
+    #[test]
+    fn clone_repo() {
```
Comment (zmitchell) [inline crates/flox-rust-sdk/src/models/floxmetav2/mod.rs:171]: nonblocking suggestion: for tests that are longer than just a few lines I like to add comments explaining what's happening in the test so it's easier to look at the test without context and understand the expected behavior.

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 3]
```diff
@@ -448,6 +473,55 @@ fn gcroots_dir(flox: &Flox, pointer: &ManagedPointer) -> PathBuf {
...
+        self.floxmeta
+            .git
+            .fetch_ref(
+                "origin",
+                &format!("{}:", remote_branch_name(&self.system, &self.pointer)),
+            )
+            .unwrap();
+
+        // Check whether we can fast-forward merge the remote branch into the local branch
+        // In not the environment has diverged.
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:489]:
```suggestion
        // If not the environment has diverged.
```

File: crates/flox-rust-sdk/src/providers/git.rs [hunk 4]
```diff
@@ -434,6 +434,20 @@ impl GitCommandProvider {
         Ok(())
     }
 
+    pub fn push2(
```
Comment (zmitchell) [inline crates/flox-rust-sdk/src/providers/git.rs:437]: blocker: surely there's a more informative name for this method? At the very least can you add a doc comment?

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 5]
```diff
@@ -448,6 +473,55 @@ fn gcroots_dir(flox: &Flox, pointer: &ManagedPointer) -> PathBuf {
...
+            .unwrap();
+
+        lock_env(
+            &self.system,
+            &self.pointer,
+            &self.floxmeta,
+            self.path.join(GENERATION_LOCK_FILENAME),
+        )
+        .unwrap();
+        Ok(())
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:521]: suggestion (blocking unless you want to do it in a separate PR): pull needs to do a build

File: crates/flox/src/commands/environment.rs [hunk 6]
```diff
@@ -582,16 +594,32 @@ pub enum PushFloxmainOrEnv {
     },
 }
 
+#[derive(Debug, Clone, Bpaf)]
+enum PullSelect {
+    New {
+        /// Directory to create the environment in (default: current directory)
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:600]: Question: Is this saying that if you don't have a `.flox` and you do `flox pull some/env` it will create the `.flox` inside of this directory, or that it will use this directory _as_ `.flox` i.e. it will put `env.json` inside of this directory?

File: crates/flox/src/commands/environment.rs [hunk 7]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
 }
 
 impl Pull {
-    pub async fn handle(self, _flox: Flox) -> Result<()> {
+    pub async fn handle(self, flox: Flox) -> Result<()> {
         subcommand_metric!("pull");
+        match self.pull_select {
+            PullSelect::New { dir, remote } => {
+                let dir = dir.unwrap_or_else(|| std::env::current_dir().unwrap());
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:635]: nonblocking suggestion: add some `debug!` statements so we can follow which type of pull the command resolved to.

File: crates/flox/src/commands/environment.rs [hunk 8]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+    /// Opens the environment and calls [ManagedEnvironment::pull] on it,
+    /// which will update the lockfile.
+    fn pull_existing_environment(
+        flox: &Flox,
+        dot_flox_path: PathBuf,
+        pointer: ManagedPointer,
+    ) -> Result<()> {
+        let mut env = ManagedEnvironment::open(flox, pointer, dot_flox_path)
+            .context("Could not open environment")?;
+        env.pull().context("Could not pull environment")?;
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:667]:
```suggestion
        env.pull().context("could not pull environment")?;
```

File: crates/flox/src/commands/environment.rs [hunk 9]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+    ///
+    /// If the directory already exists, this will fail early.
+    /// If opening the environment fails, the .flox/ directory will be cleaned up.
+    fn pull_new_environment(
+        flox: &Flox,
+        dot_flox_path: PathBuf,
+        env_ref: EnvironmentRef,
+    ) -> Result<()> {
+        if dot_flox_path.exists() {
+            bail!("Cannot pull a new environment into an existing one")
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:685]:
```suggestion
            bail!("cannot pull a new environment into an existing one")
```

File: crates/flox/src/commands/environment.rs [hunk 10]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+        }
+        let pointer = ManagedPointer::from(env_ref);
+
+        let pointer_content =
+            serde_json::to_string_pretty(&pointer).context("Could not serialize pointer")?;
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:690]:
```suggestion
            serde_json::to_string_pretty(&pointer).context("could not serialize pointer")?;
```

File: crates/flox/src/commands/environment.rs [hunk 11]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+        let pointer_path = dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME);
+
+        fs::create_dir_all(&dot_flox_path).context("Could not create .flox/ directory")?;
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:693]:
```suggestion
        fs::create_dir_all(&dot_flox_path).context("could not create .flox/ directory")?;
```

File: crates/flox/src/commands/environment.rs [hunk 12]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+        fs::write(pointer_path, pointer_content).context("Could not write pointer")?;
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:694]:
```suggestion
        fs::write(pointer_path, pointer_content).context("could not write pointer")?;
```

File: crates/flox/src/commands/environment.rs [hunk 13]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+            )) = err
+            {
+                bail!(indoc! {"
+                Could not pull environment: not logged in to floxhub.
+
+                Please login to floxhub with `flox auth login`
+                "});
+            }
+
+            fs::remove_dir_all(dot_flox_path).context("Could not clean up .flox/ directory")?;
```
Comment (zmitchell) [inline crates/flox/src/commands/environment.rs:709]:
```suggestion
            fs::remove_dir_all(dot_flox_path).context("could not clean up .flox/ directory")?;
```

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 14]
```diff
@@ -448,6 +473,55 @@ fn gcroots_dir(flox: &Flox, pointer: &ManagedPointer) -> PathBuf {
     flox.cache_dir.join("run").join(pointer.owner.to_string())
 }
 
+impl ManagedEnvironment {
+    #[allow(unused)]
+    pub fn pull(&mut self) -> Result<(), ManagedEnvironmentError> {
+        // Fetch the remote branch into FETCH_HEAD
+        self.floxmeta
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:480]: suggestion nonblocking: could you replace the fetch + branch_contains_commit + push2 with a single `fetch remote_branch_name:project_branch`?

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 15]
```diff
@@ -448,6 +473,55 @@ fn gcroots_dir(flox: &Flox, pointer: &ManagedPointer) -> PathBuf {
...
+
+        lock_env(
+            &self.system,
+            &self.pointer,
+            &self.floxmeta,
+            self.path.join(GENERATION_LOCK_FILENAME),
+        )
+        .unwrap();
+        Ok(())
+    }
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:522]: question nonblocking: are we doing force in a separate PR?

**Review by zmitchell — DISMISSED** [review]

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 16]
```diff
@@ -448,6 +473,55 @@ fn gcroots_dir(flox: &Flox, pointer: &ManagedPointer) -> PathBuf {
...
+            .push2(
+                ".",
+                format!(
+                    "FETCH_HEAD:refs/heads/{project_branch}",
+                    project_branch = branch_name(&self.system, &self.pointer, &self.path)?
+                ),
+            )
+            .unwrap();
+
+        lock_env(
```
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:514]: suggestion blocking: add a TODO to not do another fetch or change `lock_env` to `write_lockfile` or something that doesn't run another fetch

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 14 above)
Comment (mkenigs) [inline crates/flox-rust-sdk/src/models/environment/managed_environment.rs:480]:
```suggestion
        // TODO just do a build in this method if open already did the pulling and locking
        self.floxmeta
```

File: crates/flox-rust-sdk/src/providers/git.rs (hunk 4 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/providers/git.rs:437]: yes, there was a `push` method in the trait already that i had to work around. will make this clearer.

Comment (mkenigs) [inline crates/flox-rust-sdk/src/providers/git.rs:437]: suggestion nonblocking: add a comment explaining when to use push2 vs push

File: crates/flox/src/commands/environment.rs [hunk 17]
```diff
@@ -582,16 +594,32 @@ pub enum PushFloxmainOrEnv {
     },
 }
 
+#[derive(Debug, Clone, Bpaf)]
+enum PullSelect {
+    New {
+        /// Directory to create the environment in (default: current directory)
+        dir: Option<PathBuf>,
+        /// ID of the environment to pull
+        remote: EnvironmentRef,
```
Comment (mkenigs) [inline crates/flox/src/commands/environment.rs:603]: question nonblocking: should this be an optional positional? I think that's what we'll have for push which would be more consistent

File: crates/flox/src/commands/environment.rs (hunk 6 above)
Comment (ysndr) [reply crates/flox/src/commands/environment.rs:600]: it will create _a new_ `.flox` in that directory linked to `some/env`

currently we equal .flox with an environment (considering we might add flox config that might not always be true but for now is a good heuristic) so if there is a .flox present in `dir` we bail

File: crates/flox/src/commands/environment.rs [hunk 18]
```diff
@@ -600,26 +628,89 @@ pub struct Pull {
...
+        let pointer = ManagedPointer::from(env_ref);
+
+        let pointer_content =
+            serde_json::to_string_pretty(&pointer).context("Could not serialize pointer")?;
+        let pointer_path = dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME);
+
+        fs::create_dir_all(&dot_flox_path).context("Could not create .flox/ directory")?;
+        fs::write(pointer_path, pointer_content).context("Could not write pointer")?;
+
+        let result = ManagedEnvironment::open(flox, pointer, &dot_flox_path);
```
Comment (mkenigs) [inline crates/flox/src/commands/environment.rs:696]: suggestion blocking: need to build

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: crates/flox-rust-sdk/src/flox.rs [hunk 19]
```diff
@@ -66,6 +66,7 @@ pub struct Flox {
     pub uuid: uuid::Uuid,
 
     pub floxhub_token: Option<String>,
+    pub floxhub_host: String,
```
[the remark this answers is not in the document]
Comment (ysndr) [reply crates/flox-rust-sdk/src/flox.rs:69]: this is very much just temporary until we support setting other hosts.
When we do i expect that info to reside in the pointer and there being another flag/option to give flox a default.

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 1 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:415]: `lock_env_pointer`?
`lock_env_descriptor`?

proposal to change `pointer` to `descriptor` :D

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 5 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:521]: separate PR :)
image

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 14 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:480]: I'd prefer to tell explicitly if the branches diverged

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 15 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:522]: yes

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 14 above)
Comment (mkenigs) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:480]: What do you mean explicitly? `fetch` will fail if the branches have diverged unless you put `+` in the refspec

File: crates/flox/src/commands/environment.rs (hunk 17 above)
Comment (ysndr) [reply crates/flox/src/commands/environment.rs:603]: maybe, I'd like to circle back with Jenny/graham on the exact flags positionals, descriptions

**Comment by limeytexan** [comment]
> Currently missing tests as `git.hub.flox.dev` seems hardcoded at the moment and we are missing a test user to validate against (i think)

The idea as discussed (and I thought recorded in the spec) was to have a single configurable "floxhub" DNS name that would serve up the floxhub WebUI, and then the git server for that floxhub "realm" would be constructed as that name with the `git.` prefix added to the front. The default floxhub DNS name is hub.flox.dev, and the corresponding default git server would be git.hub.flox.dev.

I imagine there will be other well-known DNS names for other floxhub-associated services over time, but that git server is the first and probably the most obvious example.

File: crates/flox-rust-sdk/src/models/environment/managed_environment.rs (hunk 14 above)
Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:518]: fetch will fail also when the url is unreachable for example.
I'd like to inform someone explicitly that their environments diverged.
Showing them a raw git exception seems rather uninformative internal detail..
Unless we start parsing git errors (at least some well known ones) i guess trying to fetch the branch first is cheap and effective enough for now.

Comment (ysndr) [reply crates/flox-rust-sdk/src/models/environment/managed_environment.rs:518]: ill wait with building until we have building implemented for managed envs..

File: crates/flox/src/commands/environment.rs (hunk 18 above)
Comment (mkenigs) [reply crates/flox/src/commands/environment.rs:722]: Revisiting is fine

**Review by mkenigs — APPROVED** [review]
